dataflow: structure reduce accumulators

benesch opened the pull request:
Introduce a new "Accumulator" enumeration for processing accumulable
reductions. Previously, every aggregate accumulated into three i128
counters, even though only the Any and All aggregates actually
needed all three counters. The Accumulator enum introduced here allows
for different aggregates to use different numbers and types of counters.

This should actually result in space savings. While the enum requires an
additional tag byte, that space is more than made up for by accumulating
into isizes where possible. Unclear if the additional switch statement
in the `plus_equals` assignment has a performance cost, though.

The goal here is to make room for an `Accumulator::Decimal` variant
which accumulates the necessary fields for the new decimal type (#5274).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/materializeinc/materialize/6217)
<!-- Reviewable:end -->

frankmcsherry commented on src/dataflow/src/render/reduce.rs:
```diff
@@ -1063,6 +1066,135 @@ where
+        ///
+        /// This value is not actually observed directly by the code in this
+        /// module, but it serves to ensure that `is_zero()` returns false when
+        /// at least one non-NULL value has been observed. This is required to
```
should this be "when at least one value ..."? It seems like it is here to ensure that *null* values get noticed, rather than non-null values?

benesch replied:
Oops, yes. This text is no longer present in the latest revision though.

ruchirK commented on src/dataflow/src/render/reduce.rs:
```diff
@@ -1063,6 +1066,135 @@ where
+    }
+}
+
+impl Multiply<isize> for Accumulator {
```
I haven't looked at the new trait bounds but am a bit surprised you need multiply here because its only going in a reduce

frankmcsherry replied:
`explode` requires it. Because you might have `(record, time, 57)` and want to turn that in to something accumulable, and need to do something with the `57` (multiply it).

frankmcsherry commented on src/dataflow/src/render/reduce.rs:
```diff
@@ -1063,6 +1066,135 @@ where
+        total: isize,
+    },
+    /// Accumulates simple numeric values.
+    SimpleNumber {
+        /// The sum of all non-NULL values observed.
+        sum: i128,
+        /// The number of non-NULL values observed.
+        count: isize,
+    },
```
Does this need a `total` field? I'm just going off of the comments, but if only non-null things contribute, we'll miss null values going in to a sum and produce nothing rather than zero.

frankmcsherry replied:
Also, as they have similar sizes, would it make sense to unify this with `Count`?

frankmcsherry replied:
Reading ahead, it looks like `Datum::Null` maps to a `(0, 0)` entry for sums, which seems like it runs the risk of not producing an output for a `Sum(Null)` aggregation.

benesch replied:
Indeed. Fixed to track both `non_nulls` and `total`, as was done before.

frankmcsherry commented on src/dataflow/src/render/reduce.rs:
```diff
@@ -1063,6 +1066,135 @@ where
+}
+
+impl Default for Accumulator {
+    fn default() -> Accumulator {
+        Accumulator::Count {
+            non_nulls: 0,
+            total: 0,
+        }
+    }
```
I think this is fine because DD doesn't use `default()`, but if this accidentally gets used as a zero, it would panic the addition unless it were changed to first check `.is_zero()`. No issue that I can see, but wanted to flag.

benesch replied:
Ack, restructured with an explicit `Accumulator::Zero` type that can be added to any other type.

ruchirK commented on src/dataflow/src/render/reduce.rs:
```diff
@@ -1095,80 +1227,94 @@ where
-                    diffs[3 * accumulable_index + 2] = agg2;
+    let easy_cases = collection.inner.as_collection().explode({
+        move |(key, row)| {
+            let mut diffs = vec![Accumulator::default(); num_accumulable];
```
so I don't think its legit to do this here, because the ith accumulator may not match the default type (I'm pretty sure thats why the test is panicking)

ruchirK replied:
the intent here is if you did something like
```
sum(x), sum(distinct y), sum(z), sum(distinct a)
```

you would get 3 sets of diffs per key

one containing `[accumulator for sum(x), 0, accumulator for sum(z), 0]` i.e. one for all the non-distinct or easy cases, and then one each for the distinct cases e.g.

`[0, accumulator for sum(distinct y), 0, 0]` and `[0, 0, 0 accumulator for sum(distinct a)]` and then all 3 (or really n_distinct aggregations + 1) diffs would get merged together. i hope that makes sense

frankmcsherry replied:
I think you should be able to use `vec![]` as the default. The `Vec` implementation of `Difference` treats absent coordinates as zero, without needing a zero to populate them. I think. It's meant to, at least.

frankmcsherry replied:
Actually, I take that back I have no clue what is going on here, but ... if the panicking is happening because `default()` has variant `Count` which cannot be added to anything else, starting with an empty vector should sort it out. @ruchirK has various comments about the distinct vs non-distinct stuff, and tbh I forgot about that and don't know what is up. :D

ruchirK replied:
but then you would need to be more careful about how you populate something at an arbitrary index right? imo plus one could just check either if the rhs or lhs is_zero and in that case, forward the opposite value

ruchirK replied:
I was just explaining why its set up the way that it is -- specifically that we need to see a specific number of accumulators in each vector, and each accumulator needs to be in the right position. I _think_ based on the last time I tried to do this starting with an empty vector makes this slightly harder because then if you want to put a value at index say 3, you have to remember to also fill indexes 0, 1, and 2 with something

frankmcsherry replied:
The thing I don't know tbh is whether the correct action is an interleaving of the results or a sum. Probably an interleaving, right?

frankmcsherry replied:
Or, wait, your example suggests that it wouldn't be an interleaving, and the contributions would each have explicit zeros in them. But probably needs careful logic in either case. Seems like you should be able to look at the contributions, and at most one should be non-zero and you should select it, right? Or harden `plus_equals` at the probably negligible cost of some branch.

frankmcsherry replied:
Sorry, the "empty vector" thing was a red herring probably. I was explaining that `vec![]` is a valid argument to `plus_equals` and is treated as having empty values. Doing the vector-wise `plus_equals` could work out, except that it would inherit some zero values from the accumulations (idk where they get installed). Will stop typing without reading the surrounding code. :)

ruchirK replied:
I think this is the simple diff that fixes the test failure
```
altaria-2:materialize Test$ git diff
diff --git a/src/dataflow/src/render/reduce.rs b/src/dataflow/src/render/reduce.rs
index b1a1cbb38..52fb2853d 100644
--- a/src/dataflow/src/render/reduce.rs
+++ b/src/dataflow/src/render/reduce.rs
@@ -1123,6 +1123,15 @@ impl Semigroup for Accumulator {
     }

     fn plus_equals(&mut self, other: &Accumulator) {
+        if other.is_zero() {
+            return;
+        }
+
+        if self.is_zero() {
+            *self = other.clone();
+            return;
+        }
+
         match (self, other) {
             (
                 Accumulator::Bool {
```

and I think we actually don't even need this because we know exactly what accumulators we are going to produce and in what order, so we could instantiate a vec full of zero instances of all the accumulators and then clone that whenever we want a new "blank" diff vec if that makes sense

ruchirK replied:
its dinnertime here but i can type that up pretty easily

ruchirK replied:
As promised, I think this is the other approach one could take
```
altaria-2:materialize Test$ git diff
diff --git a/src/dataflow/src/render/reduce.rs b/src/dataflow/src/render/reduce.rs
index b1a1cbb38..eef0a105d 100644
--- a/src/dataflow/src/render/reduce.rs
+++ b/src/dataflow/src/render/reduce.rs
@@ -1210,7 +1210,7 @@ fn build_accumulable<G>(
         full_aggrs,
         simple_aggrs,
         distinct_aggrs,
-        num_accumulable,
+        num_accumulable: _,
     }: AccumulablePlan,
     prepend_key: bool,
 ) -> Arrangement<G, Row>
@@ -1228,6 +1228,22 @@ where
     // reduced if we want to specialize for the aggregations.

     let float_scale = f64::from(1 << 24);
+    // Instantiate a default vector for diffs with the correct types at each position.
+    let diffs: Vec<_> = full_aggrs
+        .iter()
+        .map(|f| match f.func {
+            AggregateFunc::Count | AggregateFunc::Dummy => Accumulator::Count {
+                non_nulls: 0,
+                total: 0,
+            },
+            AggregateFunc::Any | AggregateFunc::All => Accumulator::Bool {
+                trues: 0,
+                falses: 0,
+                total: 0,
+            },
+            _ => Accumulator::SimpleNumber { sum: 0, count: 0 },
+        })
+        .collect();

     // Two aggregation-specific values for each aggregation.
     let datum_to_accumulator = move |datum: Datum, aggr: &AggregateFunc| {
@@ -1296,8 +1312,9 @@ where
     let mut to_aggregate = Vec::new();
     // First, collect all non-distinct aggregations in one pass.
     let easy_cases = collection.inner.as_collection().explode({
+        let diffs = diffs.clone();
         move |(key, row)| {
-            let mut diffs = vec![Accumulator::default(); num_accumulable];
+            let mut diffs = diffs.clone();
             // Try to unpack only the datums we need. Unfortunately, since we
             // can't random access into a Row, we have to iterate through one by one.
             // TODO: Even though we don't have random access, we could still avoid unpacking
@@ -1320,6 +1337,7 @@ where
     // Next, collect all aggregations that require distinctness.
     for (accumulable_index, datum_index, aggr) in distinct_aggrs.into_iter() {
         let mut packer = RowPacker::new();
+        let diffs = diffs.clone();
         let collection = collection
             .map(move |(key, row)| {
                 let value = row.iter().nth(datum_index).unwrap();
@@ -1332,7 +1350,7 @@ where
             .explode({
                 move |(key, row)| {
                     let datum = row.iter().next().unwrap();
-                    let mut diffs = vec![Accumulator::default(); num_accumulable];
+                    let mut diffs = diffs.clone();
                     diffs[accumulable_index] = datum_to_accumulator(datum, &aggr.func);
                     Some((key, diffs))
                 }
```

which doesn't introduce any additional branches into `plus_equals`. Both of these diffs independently pass the sqllogictest that is currently failing in this branch.

benesch replied:
I like the look of that second diff. Done!

frankmcsherry commented on src/dataflow/src/render/reduce.rs:
```diff
@@ -1202,8 +1344,7 @@ where
 
     collection
         .arrange_by_self()
-        .reduce_abelian::<_, OrdValSpine<_, _, _, _>>(
-            "ReduceAccumulable", {
+        .reduce_abelian::<_, OrdValSpine<_, _, _, _>>("ReduceAccumulable", {
```
This is .. what like `rustfmt` changed or something?

benesch replied:
There was a string literal that was longer than 100 chars that was causing rustfmt to just give up. I split that up over two lines so now rustfmt is reformatting everything. 🤷🏽‍♂️

frankmcsherry commented on src/dataflow/src/render/reduce.rs:
```diff
@@ -1212,66 +1353,109 @@ where
                         (AggregateFunc::Dummy, _) => Datum::Dummy,
                         // Below this point, anything with only nulls should be null.
-                        (_, 0) => Datum::Null,
+                        (_, Accumulator::SimpleNumber { sum: _, count: 0 }) => Datum::Null,
```
I think this would be very surprising to see, as I think it means that we have a net-zero number of records, and a non-zero `sum` would be an error (and a zero sum should be a zero accumulation and not presented to us).

benesch replied:
Good point. This is refactored now per discussion above.

ruchirK reviewed (approved):
lgtm! thanks for jumping on this

benesch commented:
TYVM for the thorough reviews! I think I've addressed all feedback. A proper "zero diff' vector is now used that gets the types right across all accumulations, which should fix the panic and incorrect `sum` results. It also avoids the need for implementing `Default`, which should resolve some stress.

philip-stoev commented:
No wrong results with a randomized workload heavy in aggregations

ruchirK commented on src/dataflow/src/render/reduce.rs:
```diff
@@ -1212,66 +1359,114 @@ where
+                            _,
+                            Accumulator::SimpleNumber {
+                                non_nulls: 0,
+                                total: 1..=isize::MAX,
```
this match will compile down to something reasonable right? we won't be linearly scanning through this range to see if total matches it?

ruchirK replied:
tried testing this out in compiler explorer couldn't convince it to emit useful output

frankmcsherry replied:
It would be great to drop a comment in here explaining, for sure. Is it something that can be match guarded with a `total > 0` instead (which strikes me as more clear, but opinions could vary).

frankmcsherry replied:
It is a bit at adds with the logic above, where e.g. the `Any`/`All` aggregation logic don't check to see if total is positive, and in the error cases that this catches they will just produce an output. Perhaps it might make more sense to validate the data first if we want that (i.e. `total > 0`) and then decode it?

Hey related to that: the previous version of the code factored out `total` from the "variants"; would it make sense to do that here too, using a `(Accumulator, isize)` where the second coordinate always gets a `1` and can be tested for strict positivity? It seems like (not knowing future plans for decimals) that is where this is headed, and splitting it out reduces the chances that there might be varying implementations for the different variants.

ruchirK replied:
IIUC this branch is saying "if the input was all nulls and we know there was at least some input, return a null", and the equivalent logic in the boolean case would be:

```
if trues == 0 && falses == 0 && total > 0
```

which i think is implicitly caught by the two branches in the any / all cases. I'm not sure factoring out `total` would make this more clear I'm not sure what would. I also think this pr is fine to merge as is (assuming we know that the existing match is not pathological) and I'm happy to tinker and iterate on this after that?

ruchirK replied:
Just to come to what frank said earlier about a match guard:

```
diff --git a/src/dataflow/src/render/reduce.rs b/src/dataflow/src/render/reduce.rs
index 3226e1811..6d4e7baf9 100644
--- a/src/dataflow/src/render/reduce.rs
+++ b/src/dataflow/src/render/reduce.rs
@@ -1439,10 +1439,10 @@ where
                             _,
                             Accumulator::SimpleNumber {
                                 non_nulls: 0,
-                                total: 1..=isize::MAX,
+                                total,
                                 ..
                             },
-                        ) => Datum::Null,
+                        ) if *total > 0 => Datum::Null,
                         // If any non-nulls, just report the aggregate.
                         (AggregateFunc::SumInt32, Accumulator::SimpleNumber { accum, .. }) => {
                             Datum::Int64(*accum as i64)
```

compiles and probably does the right thing (haven't tested) and is maybe a simple enough change to throw in?

benesch replied:
This definitely compiles down to something reasonable! https://godbolt.org/z/rYMdndPGf

Personally I find it clearer to avoid match guards whenever something is expressible via a pattern, but clearly empirically that opinion is not widely shared. I'll update.

ruchirK replied:
omg ofc it wasn't working when I was testing it because I was using a `fn` instead of a `pub fn`!! I see now that there's a line of text at the bottom that says "Output (0/10)" indicating 0 happy lines from the compiler and 10 angry ones :D

ruchirK reviewed (commented):
other than the nit about the one match arm that seems scary to me - this seems totally great!

benesch commented:
Ok, updated to use a match guard. I'm going to take @ruchirK up on his offer and merge this as is!
